stream: Expose DuplexPair API

awwright opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

The ["Duplex Pair" pattern](https://man7.org/linux/man-pages/man2/socketpair.2.html) is a function that creates two symmetrical Duplex streams, such that what is written on one side will become readable on the other.

This pattern is used by the `tls` module in core, and in a great number of the tests. This PR merges the two implementations together into a single `stream.DuplexPair` implementation.

See #29986 for a discussion of this feature.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

mscdex commented:
Why should we be exposing this publicly? Is this a common use case? Is there a reference issue with more information?

awwright commented:
@mscdex I think it's a much more familiar way of creating a Duplex stream for another party, instead of having to us the push/_write/_read/_destroy/_final API. I think it'd be a great benefit if more people moved in the direction of creating a stream using a symmetrical API... see #29986 for a discussion; @addaleax suggested filing a PR over there.

awwright commented on test/parallel/test-http2-session-gc-while-write-scheduled.js:
```diff
@@ -14,7 +14,7 @@ const tick = require('../common/tick');
   // This creates a session and schedules a write (for the settings frame).
   let client = http2.connect('http://localhost:80', {
-    createConnection: common.mustCall(() => makeDuplexPair().clientSide)
+    createConnection: common.mustCall(() => new DuplexPair().clientSide)
```
Would anyone like to speculate why this test creates a duplex pair and then never uses the other side?

mscdex commented on lib/_stream_pair.js:
```diff
@@ -38,12 +38,13 @@ class DuplexSocket extends Duplex {
-  const clientSide = new DuplexSocket();
-  const serverSide = new DuplexSocket();
+function DuplexPair() {
+  if (!(this instanceof DuplexPair))
```
Can we avoid adding this for new APIs? We may as well use a `class`, which would force `new` anyway.

awwright replied:
Yeah I'll check it out.
Does Node.js do it this way anywhere else? I was trying too copy how the other stream classes work.

addaleax replied:
We use `class` for almost everything since `class` has been introduced into the language. Unfortunately, we cannot convert this old-style pattern to ES6 classes without a breaking change.

awwright replied:
@addaleax Would it be possible to get a deprecation notice for that, then? Or is there some documentation on what causes the breaking change? (My understanding was that `class` is just syntactic sugar.)

addaleax replied:
> @addaleax Would it be possible to get a deprecation notice for that, then?

I don’t think this is something that we’re ever realistically going to be able to change, so there’s no real reason to deprecated it, I assume.

> Or is there some documentation on what causes the breaking change? (My understanding was that `class` is just syntactic sugar.)

It’s the fact that classes *have* to be invoked with `new`, and will throw otherwise. That”s not the case in the current code here.

awwright replied:
@addaleax Thanks! I'm working on that now.

Sort of off-topic follow-up, is calling builtins without `new` discouraged too then? e.g. `const now = Date();`? I've been in discussions with a couple core members on the IRC channel, whos said more-or-less "new is obsolete these days" but I assume that was just a personal opinion around various programming styles.

addaleax replied:
@awwright `Date` is special, because `Date()` and `new Date()` do different things :)

But yes, for a function that implements a pattern like this (returns an instance of itself even when called without `new`), I would recommend using `new` because that allows transitioning to classes.

awwright replied:
I'm trying this code:
```javascript
class DuplexPair {
  [Symbol.iterator] = Array.prototype[Symbol.iterator];
  length = 2;
  constructor() {
    const clientSide = this[0] = new DuplexSide();
    const serverSide = this[1] = new DuplexSide();
    this.serverSide = clientSide[kOtherSide] = serverSide;
    this.clientSide = serverSide[kOtherSide] = clientSide;
  }
}
```

This passes the `test-release` suite, but the acorn dependency can't grok it ("SyntaxError: unexpected token (42:20)" which is at the first closing square bracket). Am I missing something here?

addaleax replied:
@awwright Yeah, we’ve run into trouble with acorn in the past for private properties as well, so I guess this falls into a similar category :/ Fwiw, I think if you do this, you could also use `class DuplexPair extends Array`, to automatically get `length` and the iterator property…

mscdex commented:
If we're going to make this public, I think we should be using more generic property names instead of 'client' and 'server' since it's no longer specific to `tls`.

awwright commented:
@mscdex I considered exposing `socket1` and `socket2`, which is what `tls` uses (the test suite uses `clientSide` and `serverSide`). I settled on additionally exposing an array called `pair` so that you can pick any name you want:

```javascript
const [ socket1, socket2 ] = new DuplexPair().pair;
```

This is what TLS now does, in the patch. Do you think that's sufficient?

Some other ideas:

* make the while object iterable, so you can do e.g. `const [ a, b ] = DuplexPair();`
* `socket1` and `socket2` (what `tls` uses)
* `upstream` and `downstream`
* `inside` and `outside`
* `a` and `b` (for brevity when doing `const { a: socket1, b: socket2 } = ...`)

I think it'd be OK to expose multiple names so applications can pick one that looks most relevant.

---

To illustrate my thought process, I think it's important to have this be a class like `Duplex` or `PassThrough` is, because I'm playing with an idea for a `SimplexPair` class, which is nothing more than a `PassThrough` with the addition of `SimplexPair#readableSide` and `SimplexPair#writableSide` properties that are only `Readable` and `Writable`, respectively.

With this, you can do patterns that maintain encapsulation, like

```javascript
function generate(){
   const stream = new SimplexPair;
   stream.write("foo");
   return stream.readableSide;
}
```

addaleax commented:
> * `socket1` and `socket2` (what `tls` uses)

I like this because it matches `port1` and `port2` from `new MessageChannel()`. (Alternatively: `stream1` and `stream2`.)

> * make the while object iterable, so you can do e.g. `const [ a, b ] = DuplexPair();`

Fwiw, that’s not exclusive with any naming choices. I like it. :+1:

mscdex commented:
I'd prefer something even more generic than `socket1` and `socket2` because this interface doesn't utilize `net` in any way. Maybe something as simple as `side1` and `side2`?

awwright commented:
Check my work on the iterable here:

```javascript
function DuplexPair() {
  if (!(this instanceof DuplexPair))
    return new DuplexPair();
  const clientSide = this[0] = new DuplexSide();
  const serverSide = this[1] = new DuplexSide();
  this.length = 2;
  this[Symbol.iterator] = Array.prototype.iterator;
  clientSide[kOtherSide] = serverSide;
  serverSide[kOtherSide] = clientSide;
}
```

afaict, this should behave the same as an Array, and still permit iterating over the other keys in a for..in loop.

I like side1 and side2 better than anything else, except I have this hesitation because the numbers don't line up with the array indices, and I'm slightly personally attached to `serverSide`/`clientSide`, since a lot of applications will fall into this paradigm.

Do we need the named properties at all, though?

If performance with the iterator function is a concern, you can still do

```
const { 0: side1, 1: side2 } = new DuplexPair;
```

How does this sound?

awwright commented on lib/tls.js:
```diff
@@ -44,7 +44,7 @@ const { getRootCertificates, getSSLCiphers } = internalBinding('crypto');
 const EventEmitter = require('events');
 const { URL } = require('internal/url');
-const DuplexPair = require('internal/streams/duplexpair');
+const { DuplexPair } = require('stream');
```
Should this refer to _stream_duplexpair instead?

addaleax replied:
All the underscore-prefixed modules are considered legacy, and we definitely want to avoid introducing new ones or their usage

addaleax commented on lib/_stream_duplexpair.js:
```diff
@@ -1,15 +1,12 @@
 'use strict';
```
Can you undo the file rename?

awwright replied:
@addaleax You mean you like `_stream_pair.js` better? I figured this is more consistent with `lib/internal/streams/duplexpair.js` and with the name of the property exported by "stream"

awwright replied:
Oh, so keep the stream under lib/internal, that makes sense

addaleax commented on lib/_stream_duplexpair.js:
```diff
@@ -1,15 +1,12 @@
-} = primordials;
-
 const { Duplex } = require('stream');
+const assert = require('assert');
```
```suggestion
const assert = require('internal/assert');
```

addaleax commented:
@nodejs/streams Any further thoughts on this?

mscdex commented on doc/api/stream.md:
````diff
@@ -1508,6 +1510,25 @@ unless `emitClose` is set in false.
+```
+
+Whatever is written to one stream is made readable on the other. It provides
+behavior analagous to a network connection, where the data which a client writes
````
```suggestion
behavior analogous to a network connection, where the data which a client writes
```

ronag reviewed (dismissed):
I don't think it's a good idea to expose this. I don't feel it is motivated enough. Why do we need this? Also my [concerns](https://github.com/nodejs/node/issues/29986#issuecomment-549126956) in the original issue remains.

I've probably missed this somewhere, but what exactly is the significant advantage over just using `Duplex`?

EDIT: Just to clarify. Conceptually I do agree this might be a good idea. I just don't feel it makes a significant enough difference that it's worth to make public and maintain, i.e. just using `Duplex` should work and be more performant (?). Also I'm not entirely comfortable with the implementation. There are already [plenty of edge cases](https://github.com/nodejs/node/issues/34108) in the public streams API's that should be resolved and I would like to avoid adding more surface to that list. As mentioned in the [issue](https://github.com/nodejs/node/issues/29986#issuecomment-549126956) there are edge cases with this as well that might or might not need to be resolved.

awwright commented:
@ronag Those are great questions, and I've been making a list of my own that I'll try fix up and share soon here... I'd like to help identify some fixes too.

---

I'll post more detail in the Issue soon, but in a nutshell, the aim is to _simplify_ how streams are used by _reducing_ the API surface that userspace has to worry about.

The way I see streams, Writable and Readable are interfaces, not concrete implementations. (Maybe let's call the interfaces themselves `Sink` and `Source`, respectively, to differentiate them from the Node.js implementation that includes a buffer) There is a "Sink" interface, which is used to send data into a stream, and a "Source" interface, which is used to read the data off the stream. And these can be implemented any number of ways as long as it always has the same behavior (so long as `Sink#cork` prevents writes from having any affect, etc).

You might say "ensuring that the implementations ALWAYS have the same behavior would be very difficult"... and that would be correct. Fortunately, I figure there's really only two implementations you'll ever need:

1. You have a buffer (a `BufferList` in Node.js terms); writing to the Sink interface fills the buffer, and reading from the Source interface drains the buffer.

2. You have a Source that generates data as it is requested (e.g. a random number generator, or `/dev/zero`), as fast as the application can read from it.

(There's no third "A Sink interface that processes the data as fast as its written" because this is exactly what the "readable" or "data" events do.)

So, I figure Node.js implements (2) with `Readable`, but lacks an implementation of (1).

---

You bring up a good point, I would like to make sure that error handling and things can be changed or added without breaking userland stuff. If we're going to add this, let's do it the right way.

But I do think adding this in some capacity is warranted, and I don't think it adds any maintenance burden: this is already used by `tls` and most of the test suite—so it's part of the contract already (if only implicitly, in the fine print, so to speak).

ronag commented:
> used by reducing the API surface that userspace has to worry about.

But it increases the API surfaces that we have to worry about. Also, it's already confusing for users the number of different stream API's we have e.g. the different readable modes.

I'm still confused on how this is significantly different than `Duplex`.

```js
const duplex = new Duplex() 
return {
  sink: duplex,
  source: duplex
}
```

ronag commented:
@jasnell Do you make use of `DuplexPair` in your current work on e.g. quick?

mcollina commented:
If this lands it should be marked as "experimental", so we could avoid adhering to semver and ship breaking changes as fixes.

ronag commented:
I would also like to note that `DuplexPair` is only used in core by `createSecurePair` which is deprecated. So basically there isn't much motivation for it to be in core other than legacy compat.

addaleax commented:
@ronag I think the fact that we use this somewhat extensively in our tests shows that there is usefulness for this in core. We’re not going to remove this entirely from our codebase (except e.g. if we start vendoring it from an npm module at some point).

> > since this is already used by tls and most of the test suite
> 
> This is not a good argument though. tls is not well maintained as it is (not many people are comfortable with touching that code, which kinds of makes it a counterpoint) and tests are not something we really maintain per se, once they test what we need we leave them be.

We do keep using it for new tests, and across a nontrivial number of tests. In an ideal world, most of our HTTP, HTTP/2 and TLS tests would use this instead of plain network connections.

> I'm still unsure on how this is significantly different than `Duplex`.
 
It is … an entirely different thing? That comparison doesn’t really make sense, I’m afraid. If you want to compare it to anything, you could compare it to a pair `PassThrough` streams. But even then, for us to provide something out-of-the-box that would allow implementing an equivalent to this without building custom streams, we would need some way of creating `Duplex` streams that wrap around a pair of `Readable` and `Writable` streams. (That would be an alternative API that we could implement.)

It’s definitely a primitive for streams that we’re currently lacking. That can be fine, it can be something we expect userland libraries to implement, but it should be there in *some* way.

ronag commented:
>  I think the fact that we use this somewhat extensively in our tests shows that there is usefulness for this in core.

It is a bit for a special use case though? I'm not sure it's generally useful?

> We do keep using it for new tests, and across a nontrivial number of tests. In an ideal world, most of our HTTP, HTTP/2 and TLS tests would use this instead of plain network connections.

Yes, agreed. For our test suite it kind of makes sense.

> It is … an entirely different thing? That comparison doesn’t really make sense, I’m afraid. If you want to compare it to anything, you could compare it to a pair PassThrough streams. But even then, for us to provide something out-of-the-box that would allow implementing an equivalent to this without building custom streams, we would need some way of creating Duplex streams that wrap around a pair of Readable and Writable streams. (That would be an alternative API that we could implement.)
> 
> It’s definitely a primitive for streams that we’re currently lacking. That can be fine, it can be something we expect userland libraries to implement, but it should be there in some way.

This is where I get lost. In what way exactly is it a different thing? I just don't see what this is and why we need it for anything than the special case of our tests where we want to emulate a client & server communication. Even there I'm questioning if this is the actual API that is needed, i.e. what we want is a Writable/Readable pair, not a Duplex pair, which can be implemented as a duplex already, it's just that we would separate out the two different interfaces.

I feel I'm missing something fundamental here. I don't see why we need this. Maybe improving the description and examples in the docs could help me understand? Maybe an example to have as a basis for discussion would be useful?

jasnell commented:
We do not currently use it directly in the quic implementation but we implement the same basic concept in the udp pair utility. It's entirely possible that we could combine the two in the future somehow. I'm fine with adding this to core, as it is a useful utility, however, I'd be just as fine with it existing as a separate utility module published to npm. (In fact, I think a large number of our common test utilities could be done that way also and just vendored in).

ronag commented:
> We do not currently use it directly in the quic implementation but we implement the same basic concept in the udp pair utility.

But this utility is also just used for tests?

jasnell commented:
Yes, only in tests at the moment

mcollina commented:
> I would also like to note that DuplexPair is only used in core by tls.createSecurePair which is runtime deprecated since Node 8. Might even be time to consider EOL on that. So basically there isn't much motivation at the moment for it to be in core other than legacy compat.

I think we should ship an ecosystem module with DuplexPair in it, and drop `tls.createSecurePair` in v15.

ronag commented:
> I think we should ship an ecosystem module with DuplexPair in it

What is an "ecosystem module"? Something like https://www.npmjs.com/package/duplexpair?

mcollina commented:
Essentially updating that and having it in the nodejs org. cc @addaleax

jasnell commented:
@mcollina:
> I think we should ship an ecosystem module with DuplexPair in it, and drop tls.createSecurePair in v15.

I'm good with this approach also, with the caveat... deprecate in v15 and eol in v16

mcollina commented:
> I'm good with this approach also, with the caveat... deprecate in v15 and eol in v16

@jasnell it has bee runtime deprecated since v8. It's fine to remove in v15

jasnell commented:
> it has bee runtime deprecated since v8. It's fine to remove in v15

Ah, that's right. Thanks for the reminder :-)

awwright commented:
@ronag Allow me some time to put together a better argument; but (1) I concur with @addaleax that the fact this is the utility of choice in the tests, rather than using vanilla Duplex, is a strong indicator that this is a better API.

And (2) the reason I say this does not increase the API surface is because this only uses features that we promise to userland somewhere else.

I'm preparing a better argument for why this is appropriate for core, so please give me a little bit on that :)

addaleax commented:
If y’all want to discuss removing `tls.SecurePair`, I’d suggest doing that in a separate issue. I, personally, would lean towards un-deprecating it, given that the introduction of `DuplexPair` essentially removed all the maintenance burden behind it (which is also a strong argument for this PR imo).

addaleax commented:
> > I think the fact that we use this somewhat extensively in our tests shows that there is usefulness for this in core.
> 
> It is a bit for a special use case though? I'm not sure it's generally useful?

It is the Node.js equivalent of `socketpair(2)`, a concept that has existed and has been useful in UNIX programming for quite a while (especially in the single-direction form of `pipe(2)`).

> 
> > It is … an entirely different thing? That comparison doesn’t really make sense, I’m afraid. If you want to compare it to anything, you could compare it to a pair PassThrough streams. But even then, for us to provide something out-of-the-box that would allow implementing an equivalent to this without building custom streams, we would need some way of creating Duplex streams that wrap around a pair of Readable and Writable streams. (That would be an alternative API that we could implement.)
> > It’s definitely a primitive for streams that we’re currently lacking. That can be fine, it can be something we expect userland libraries to implement, but it should be there in some way.
> 
> This is where I get lost. In what way exactly is it a significantly different thing? I just don't see what "this" is and why we need it for anything than the special case of our tests where we want to emulate a client & server communication. Even there I'm questioning if this is the actual API that is needed (not that it matters, duplexpair does the job), i.e. isn't what we want a Writable/Readable pair, not a Duplex pair, which can be implemented as a duplex already, it's just that we would separate out the two different interfaces (which could be two references to a Duplex instance).

If you want something to act like a connection, then you’ll need two Writable/Readable pairs. As I mentioned above, another alternative would be adding a function that creates a `Duplex` stream from a Readable/Writable pair – in that case, using 2 `PassThrough` streams and passing them to that function would allow creating a `Duplex` pair without writing custom streams.

Basically, what I’m thinking here is that `DuplexPair` falls into the same category as `PassThrough` – there’s no real reason for Node.js core to provide `PassThrough` other than that it makes sense to have it in the basic streams primitives toolbox, and the same goes for `DuplexPair`.

ronag commented:
> a concept that has existed and has been useful in UNIX programming for quite a while

This does not necessarily mean it's a good idea? Or a good idea for Node. I'm not that much in the unix world but how commonly is this used and for what?

> (especially in the single-direction form of pipe(2)).

Isn't this basically a Duplex?

> there’s no real reason for Node.js core to provide PassThrough 

Which is not a good motivation to add another thing that there is no real reason for?

I feel this is going around in circles and I'm just repeating the same things. I don't see why we should add this without an actual example of when/how/who would use this. I feel the motivations so far have been more conceptual than practical. Maybe I'm just missing something. I don't think I have anything further useful to contribute here.

I'm removing my request for changes in order to avoid blocking this. But I ask that you consider whether we really need this.

awwright commented:
@addaleax made a much better and much simpler argument for inclusion than I was going to make; which is "It's just a missing `Duplex` variant of `PassThrough`".

@ronag I know you'd like to see evidence that this would be well used and worth the maintenance burden, but that's quite subjective, and it's difficult for me to prove well. All I have is: this is likely to be picked up in stream-related unit tests, and it does not use any internal APIs, which suggests to me a low maintenance burden, and that the benefits outweigh the costs. Maybe if there were a more objective criteria, I could help better.

As further evidence this is a widely used pattern, suitable for standardization in Node.js, here's other languages that implement similar patterns standard:

- [POSIX 2008, Linux](https://man7.org/linux/man-pages/man2/socketpair.2.html)
- [PHP](https://www.php.net/manual/en/function.stream-socket-pair.php)
- [Ruby](https://ruby-doc.org/stdlib-2.7.1/libdoc/socket/rdoc/Socket.html#method-c-pair)
- [Python](https://docs.python.org/3.9/library/socket.html#socket.socketpair)
- [Haskell](https://hackage.haskell.org/package/network-3.1.1.1/docs/Network-Socket.html#v:socketPair)
- [Go](https://golang.org/pkg/syscall/#Socketpair)
- [Rust](https://doc.rust-lang.org/std/os/unix/net/struct.UnixStream.html#method.pair)

Obviously there's some implementation differences. Many of these only work for network-type streams (for platforms that make a distinction), and are not compatible with seeking streams (like for files). Some provide true anonymous network sockets, so that the stream can still pass data between a process and its fork. Except for this one case, this PR appears to do the least required work that provides equivalent functionality.

ronag commented:
> As further evidence this is a widely used pattern, suitable for standardization in Node.js, here's other languages that implement similar patterns standard:

These are all just wrappers for https://man7.org/linux/man-pages/man2/socketpair.2.html. Which is a different thing.

Anyway, if you guys feel strongly about it and want to maintain it, go for it.

mcollina commented on lib/internal/streams/duplexpair.js:
```diff
@@ -39,13 +38,13 @@ class DuplexSocket extends Duplex {
 }
 
-class DuplexPair {
+class DuplexPair extends Array {
```
I would prefer if this was not a class a tall, but rather a factory and it did not inherit from `Array`.

mcollina replied:
@awwright you might have missed this.

mcollina commented on doc/api/stream.md:
```diff
@@ -1508,6 +1510,25 @@ unless `emitClose` is set in false.
+-->
+
+The utility class `DuplexPair` returns an iterable object with two `Duplex`
+streams on properties `0` and `1`, each connected to the other side:
```
I would document that this returns an array.

awwright commented:
@mcollina Oops, I drafted a comment but never posted it :|

Do you figure this is more consistent with the rest of the API?

@addaleax @mscdex Does a function `duplexPair` sound OK instead of the class?

addaleax commented:
@awwright Yes, that sounds good as well – Ultimately, the two are equivalent (which is also why I’m not really understanding where @mcollina’s objection is coming from, but, :woman_shrugging:).

mcollina commented:
@addaleax I think having `class DuplexPair extend Array` might be the source of future problems.

jasnell commented:
I also tend to prefer the `const [a, b] = duplexPair()` syntax here also.

It would also be worth allowing for an `options` argument on the `duplexPair()` (whether it's a function or class) that allows initial options to be set for each of the Duplex's (e.g. high water mark, default encoding, capture rejections, etc).

awwright commented:
Alright I changed this to a function and added an `options` argument... This is currently failing for me (macOS), but it appears to be a commit between 09476f71c7 to 89453cfc082304d847f74701d8b70754545244d0 inclusive. And I have to document the `options` argument better anyways.

mcollina commented on test/parallel/test-stream-duplexpair.js:
```diff
@@ -0,0 +1,94 @@
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
```
I don't think this header is needed, it's a new file and not copied from somewhere else.

awwright replied:
Alright I just took out the entire thing

awwright commented:
I just pushed out an update to the documentation describing the `options` argument. The same value is passed to both constructors.

I have one more idea: Would it be a good idea to set the "other side" reference as a private property (`this.#otherSide`)? This would benefit encapsulation, and might be more difficult to implement later on.

mcollina commented:
I would keep things as they are right now.

Mifrill commented on lib/internal/streams/duplexpair.js:
```diff
@@ -39,13 +38,11 @@ class DuplexSocket extends Duplex {
+function duplexPair(options) {
+  const side0 = new DuplexSide(options);
+  const side1 = new DuplexSide(options);
+  side0[kOtherSide] = side1;
+  side1[kOtherSide] = side0;
```
I think it's better to restrict direct access for those properties by encapsulation, so they can't be modified directly from outside the class.

awwright replied:
I agree with this... I may have been under the impression that you can't acquire the value of kOtherSide from outside the file (but actually you can enumerate these). Would you suggest a WeakMap instead?

Mifrill replied:
> Would you suggest a WeakMap instead?

well, it would be a bit better with WeakMap, something like that:
```js
const privateData = new WeakMap();

class DuplexSide extends Duplex {
  constructor() {
    privateData.set(this, {
      [kCallback]: null,
      [kOtherSide]: null,
    });
  }

  setOtherSide(otherSide) {
    const data = privateData.get(this);
    if (data[kOtherSide]) {
      throw new Error('readonly')
    }
    data[kOtherSide] = otherSide;
  }
}
```
however, even that approach cannot guarantee truly private properties that can't be directly accessed or modified, I would say it's only hope on convention. WeakMap provide a level of privacy and make it difficult for external code to accidentally interfere, but they do not make the properties completely immutable.

I suggest using the redefinition of `set` method together with `Proxy` object for that, to make sure of immutable:
```js
const handler = {
  set(target, prop, value) {
    if (prop in target && target[prop] !== null) {
      throw new Error('Property is already set.')
    } else {
      target[prop] = value
      return true
    }
  }
}

class DuplexSocket {
  constructor() {
    this[kCallback] = null
    this[kOtherSide] = null
    return new Proxy(this, handler)
  }
}
```
🗒️ `target[prop] !== null` condition needs to handle first assign properties as null in the constructor.

📓 specs:
```js
  test('allows assign kCallback.', async () => {
    const socket = new DuplexSocket()
    socket[kCallback] = 'value kCallback 1'
    expect(socket[kCallback]).toStrictEqual('value kCallback 1')
  })

  test('forbids reassign kCallback.', async () => {
    const socket = new DuplexSocket()
    socket[kCallback] = 'value kCallback 1'
    expect(() => {
      socket[kCallback] = 'value kCallback 2'
    }).toThrow(new Error('Property is already set.'))
    expect(socket[kCallback]).toStrictEqual('value kCallback 1')
  })

  test('allows assign kOtherSide.', async () => {
    const socket = new DuplexSocket()
    socket[kOtherSide] = 'value kOtherSide 1'
    expect(socket[kOtherSide]).toStrictEqual('value kOtherSide 1')
  })

  test('forbids reassign kOtherSide.', async () => {
    const socket = new DuplexSocket()
    socket[kCallback] = 'value kOtherSide 1'
    expect(() => {
      socket[kCallback] = 'value kOtherSide 2'
    }).toThrow(new Error('Property is already set.'))
    expect(socket[kCallback]).toStrictEqual('value kOtherSide 1')
  })
```

mcollina replied:
Let's not use WeakMaps for this, they are incredibly expensive from a GC standpoint.

mcollina reviewed (approved):
lgtm

mcollina commented:
@ronag could you take another look?

mcollina commented:
Can you rebase on top of main?

ronag reviewed (commented):
I still don't think we should expose this. Not blocking but I don't see the use case for it. Is there a heavily used npm package that does this? For me this is more of a slow anti pattern that we shouldn't be encouraging.

Isn't SecurePair basically deprecated?

jasnell commented:
@ronag ... since you indicated that your objection is non-blocking would you mind clearing your "Request Changes"?

ronag reviewed (commented):
Don't think it's a good idea but won't block.

jasnell commented:
> Don't think it's a good idea but won't block.

Understood. This requires a rebase anyway so there's still time to discuss and consider. @mcollina, do you have any further thoughts?

awwright commented:
Rebased; made a few minor changes to appease updated linter standards; and I moved the otherSide reference into a private property, which is a pattern that seems to be used elsewhere Node.js.

awwright commented on doc/api/stream.md:
```diff
@@ -2620,6 +2622,30 @@ unless `emitClose` is set in false.
+#### `stream.duplexPair([options])`
+
+<!-- YAML
+added: REPLACEME
```
Is this correct?

mcollina commented:
The linter is failing, can you check?

awwright commented:
@mcollina `make test` passes for me and I don't have access to three of the failed tests. Is it related to the comment I [dropped here](https://github.com/nodejs/node/pull/34111#pullrequestreview-1738424743)?

mcollina commented:
@awwright you should run the linter locally with `make lint`. It will show you the error.

devsnek commented on doc/api/stream.md:
````diff
@@ -2624,6 +2627,30 @@ unless `emitClose` is set in false.
+each being a `Duplex` stream connected to the other side:
+
+```js
+const [ clientSide, serverSide ] = duplexPair();
````
nit: perhaps something like this, so it less implies that there is anything special about the order?
```suggestion
const [ sideA, sideB ] = duplexPair();
```

aduh95 commented:
Can you please rebase and address the linter error?

awwright commented:
@aduh95 Is this good?

mcollina reviewed (approved):
lgtm

awwright commented:
I could have sworn all the tests passed when I pushed this, is that blocking the merge?

mcollina commented:
> I could have sworn all the tests passed when I pushed this, is that blocking the merge?

Yes.

```
---
duration_ms: 266.324
exitcode: 1
severity: fail
stack: |2-

  node:internal/event_target:1090
    process.nextTick(() => { throw err; });
                             ^
  AssertionError [ERR_ASSERTION]: These builtins are now unexpectedly loaded at run time.
  If this is intentional, add them to `expected.atRunTime`.

  # Note: loading more builtins at run time can lead to startup performance regression.
  - Consider lazy loading builtins that are not used universally.

  --- These could be added to expected.atRunTime ---
  [
    'NativeModule internal/streams/duplexpair'
  ]

      at Object.<anonymous> (/home/iojs/build/workspace/node-test-commit-custom-suites-freestyle/test/parallel/test-bootstrap-modules.js:242:8)
      at Module._compile (node:internal/modules/cjs/loader:1458:14)
      at Object..js (node:internal/modules/cjs/loader:1542:10)
      at Module.load (node:internal/modules/cjs/loader:1282:32)
      at Function._load (node:internal/modules/cjs/loader:1098:12)
      at TracingChannel.traceSync (node:diagnostics_channel:315:14)
      at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)
      at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:158:5)
      at MessagePort.<anonymous> (node:internal/main/worker_thread:181:26)
      at [nodejs.internal.kHybridDispatch] (node:internal/event_target:816:20) {
    generatedMessage: false,
    code: 'ERR_ASSERTION',
    actual: 7,
    expected: 0,
    operator: 'strictEqual'
  }

  Node.js v23.0.0-pre
...
```

This is relevant: https://ci.nodejs.org/job/node-test-commit-custom-suites-freestyle/36865/testReport/junit/(root)/parallel/test_bootstrap_modules/

mcollina reviewed (approved):
lgtm
